File: built-in plugins dsp/mb-filter-state-var.cpp
/**
 * MB State Variable Filter
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Multi-mode state variable filter with continuous morphing
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_FILTER_STATE_VAR_H
#define MB_FILTER_STATE_VAR_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbFilterStateVar : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-filter-state-var";
    static constexpr const char* PLUGIN_NAME    = "MB State Variable Filter";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float cutoff = 3000f;  // range [20, 20000]
    float resonance = 0.3f;  // range [0, 1]
    float morph = 0f;  // range [0, 1]
    float mix = 1f;  // range [0, 1]
    };

    MbFilterStateVar() = default;
    ~MbFilterStateVar() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.cutoff = std::clamp(params.cutoff, 20f, 20000f);
        params.resonance = std::clamp(params.resonance, 0f, 1f);
        params.morph = std::clamp(params.morph, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB State Variable Filter
        return input;
    }
};

#endif // MB_FILTER_STATE_VAR_H

File: built-in plugins dsp/mb-filter-state-var.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbFilterStateVarPlugin: PluginDefinition = { id: 'mb-filter-state-var', slug: 'mb-filter-state-var', name: 'MB State Variable Filter', category: 'effect', type: 'eq' as any, version: '1.0.0', description: 'Multi-mode state variable filter with continuous morphing', author: 'Max Booster', grade: 'A', parameters: [{ id: 'cutoff', name: 'Cutoff', type: 'float', defaultValue: 3000, minValue: 20, maxValue: 20000, automatable: true }, { id: 'resonance', name: 'Resonance', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }, { id: 'morph', name: 'LP/BP/HP Morph', type: 'float', defaultValue: 0, minValue: 0, maxValue: 1, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { cutoff: 3000, resonance: 0.3, morph: 0, mix: 1 } };

export default MbFilterStateVarPlugin;
